src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -74,6 +74,42 @@ ssize_t show_kds_custat_raw(struct kds_sched *kds, char *buf)
 	return sz;
 }
 
+/* Each line is a PS kernel, format:
+ * "idx kernel_name status usage"
+ */
+ssize_t show_kds_scustat_raw(struct kds_sched *kds, char *buf)
+{
+	struct kds_scu_mgmt *scu_mgmt = &kds->scu_mgmt;
+	char *cu_fmt = "%d,%s,0x%x,%u\n";
+	ssize_t sz = 0;
+	int i;
+
+	/* TODO: The number of PS kernel could be 64 or even more.
+	 * Sysfs has PAGE_SIZE limit, which keep bother us in old KDS.
+	 * In 128 PS kernels case, each line is average 32 bytes.
+	 * The kernel name is no more than 19 bytes.
+	 *
+	 * Old KDS shows FPGA Kernel and PS kernel in one file.
+	 * So, this separate kds_scustat_raw is better.
+	 *
+	 * But in the worst case, this is still not good enough.
+	 */
+	mutex_lock(&scu_mgmt->lock);
+	for (i = 0; i < scu_mgmt->num_cus; ++i) {
+		sz += scnprintf(buf+sz, PAGE_SIZE - sz, cu_fmt, i,
+				scu_mgmt->name[i], scu_mgmt->status[i],
+				scu_mgmt->usage[i]);
+	}
```

Comment:
If we overflow, the memory will already be corrupted, right?

Reply:
Buf would not over flow. The second argument of scnprintf() is the maximum bytes can be written.
Use scnprintf() is a very simple implementation of a device attribute (see https://github.com/torvalds/linux/blob/master/Documentation/filesystems/sysfs.rst)

BTW, this code cannot show more than PAGE_SIZE. This solution doesn't work well when we have 128 long name PS kernels and with large usage value.

Reply:
> Buf would not over flow. The second argument of scnprintf() is the maximum bytes can be written.
> Use scnprintf() is a very simple implementation of a device attribute (see https://github.com/torvalds/linux/blob/master/Documentation/filesystems/sysfs.rst)
> 
> BTW, this code cannot show more than PAGE_SIZE. This solution doesn't work well when we have 128 long name PS kernels and with large usage value.

OK. Sounds good.